File: spark/src/test/scala/geotrellis/spark/testfiles/TestFiles.scala
package geotrellis.spark.testfiles

import com.github.nscala_time.time.Imports._
import geotrellis.spark.tiling._
import geotrellis.raster.{GridBounds, TileLayout, FloatConstantNoDataCellType}
import geotrellis.spark.tiling.{LayoutDefinition, MapKeyTransform}
import org.apache.spark._
import geotrellis.spark._
import geotrellis.proj4._
import org.joda.time.DateTime

object TestFiles extends Logging {
  val ZOOM_LEVEL = 8
  val partitionCount = 4

  def generateSpatial(layerName: String)(implicit sc: SparkContext): RasterRDD[GridKey] = {
    val md = {
      val cellType = FloatConstantNoDataCellType
      val crs = LatLng
      val tileLayout = TileLayout(8, 8, 3, 4)
      val mapTransform = MapKeyTransform(crs, tileLayout.layoutDimensions)
      val gridBounds = GridBounds(1, 1, 6, 7)
      val extent = mapTransform(gridBounds)
      val keyBounds = KeyBounds(GridKey(1,1), GridKey(6,7))
      RasterMetadata(cellType, LayoutDefinition(crs.worldExtent, tileLayout), extent, crs, keyBounds)
    }

    val gridBounds = md.gridBounds
    val tileLayout = md.tileLayout

    val gridKeyTestFile = layerName match {
      case "all-ones" => new ConstantGridKeyTiles (tileLayout, 1)
      case "all-twos" => new ConstantGridKeyTiles (tileLayout, 2)
      case "all-hundreds" => new ConstantGridKeyTiles (tileLayout, 100)
      case "increasing" => new IncreasingGridKeyTiles (tileLayout, gridBounds)
      case "decreasing" => new DecreasingGridKeyTiles (tileLayout, gridBounds)
      case "every-other-undefined" => new EveryOtherGridKeyTiles (tileLayout, gridBounds, Double.NaN, 0.0)
      case "every-other-0.99-else-1.01" => new EveryOtherGridKeyTiles (tileLayout, gridBounds, 0.99, 1.01)
      case "every-other-1-else-1" => new EveryOtherGridKeyTiles (tileLayout, gridBounds, - 1, 1)
      case "mod-10000" => new ModGridKeyTiles (tileLayout, gridBounds, 10000)
    }

    val tiles =
      for(
        row <- gridBounds.rowMin to gridBounds.rowMax;
        col <- gridBounds.colMin to gridBounds.colMax
      ) yield {
        val key = GridKey(col, row)
        val tile = gridKeyTestFile(key)
        (key, tile)
      }

    new ContextRDD(sc.parallelize(tiles, partitionCount), md)
  }

  def generateSpaceTime(layerName: String)(implicit sc: SparkContext): RasterRDD[GridTimeKey] = {
    val times =
      (0 to 4).map(i => new DateTime(2010 + i, 1, 1, 0, 0, 0, DateTimeZone.UTC)).toArray


    val md = {
      val cellType = FloatConstantNoDataCellType
      val crs = LatLng
      val tileLayout = TileLayout(8, 8, 3, 4)
      val mapTransform = MapKeyTransform(crs, tileLayout.layoutDimensions)
      val gridBounds = GridBounds(1, 1, 6, 7)
      val extent = mapTransform(gridBounds)
      val keyBounds = KeyBounds(GridTimeKey(1,1,times.min), GridTimeKey(6,7, times.max))
      RasterMetadata(cellType, LayoutDefinition(crs.worldExtent, tileLayout), extent, crs, keyBounds)
    }

    val gridBounds = md.gridBounds
    val tileLayout = md.tileLayout

    val gridTimeKeyTestTiles = layerName match {
      case "gridtimekey-all-ones" => new ConstantGridTimeKeyTestTiles(tileLayout, 1)
      case "gridtimekey-all-twos" => new ConstantGridTimeKeyTestTiles(tileLayout, 2)
      case "gridtimekey-all-hundreds" => new ConstantGridTimeKeyTestTiles(tileLayout, 100)
      case "gridtimekey-coordinates" => new CoordinateGridTimeKeyTestTiles(tileLayout)
    }

    val tiles =
      for(
        row <- gridBounds.rowMin to gridBounds.rowMax;
        col <- gridBounds.colMin to gridBounds.colMax;
        (time, timeIndex) <- times.zipWithIndex
      ) yield {
        val key = GridTimeKey(col, row, time)
        val tile = gridTimeKeyTestTiles(key, timeIndex)
        (key, tile)

      }

    new ContextRDD(sc.parallelize(tiles, partitionCount), md)
  }
}

trait TestFiles { self: TestEnvironment =>
  def gridKeyTestFile(name: String) = TestFiles.generateSpatial(name)

  def gridTimeKeyTestFile(name: String) = TestFiles.generateSpaceTime(name)

  lazy val AllOnesTestFile =
    gridKeyTestFile("all-ones")

  lazy val AllTwosTestFile =
    gridKeyTestFile("all-twos")

  lazy val AllHundredsTestFile =
    gridKeyTestFile("all-hundreds")

  lazy val IncreasingTestFile =
    gridKeyTestFile("increasing")

  lazy val DecreasingTestFile =
    gridKeyTestFile("decreasing")

  lazy val EveryOtherUndefinedTestFile =
    gridKeyTestFile("every-other-undefined")

  lazy val EveryOther0Point99Else1Point01TestFile =
    gridKeyTestFile("every-other-0.99-else-1.01")

  lazy val EveryOther1ElseMinus1TestFile =
    gridKeyTestFile("every-other-1-else-1")

  lazy val Mod10000TestFile =
    gridKeyTestFile("mod-10000")

  lazy val AllOnesSpaceTime =
    gridTimeKeyTestFile("gridtimekey-all-ones")

  lazy val AllTwosSpaceTime =
    gridTimeKeyTestFile("gridtimekey-all-twos")

  lazy val AllHundredsSpaceTime =
    gridTimeKeyTestFile("gridtimekey-all-hundreds")

  /** Coordinates are CCC,RRR.TTT where C = column, R = row, T = time (year in 2010 + T).
    * So 34,025,004 would represent col 34, row 25, year 2014
    */
  lazy val CoordinateSpaceTime =
    gridTimeKeyTestFile("gridtimekey-coordinates")
}

File: geotrellis-python/tests/geotrellis/spark/testfiles/TestFiles.py
from tests.geotrellis.spark.TestEnvironment import _TestEnvironment
from geotrellis.raster.GridBounds import GridBounds
from geotrellis.raster.TileLayout import TileLayout

class _TestFiles(_TestEnvironment):

    ZOOM_LEVEL = 8
    partitionCount = 4

    @staticmethod
    def generateSpatial(layerName, sc):
        gridBounds = GridBounds(1,1,6,7)
        tileLayout = TileLayout(8,8,3,4)
        def generateSpatialTestFile():
            if layerName == "all-ones":
                return ConstantSpatialTiles (tileLayout, 1)
            elif layerName == "all-twos":
                return ConstantSpatialTiles (tileLayout, 2)
            elif layerName == "all-hundreds":
                return ConstantSpatialTiles (tileLayout, 100)
            elif layerName == "increasing":
                return IncreasingSpatialTiles (tileLayout, gridBounds)
            elif layerName == "decreasing":
                return DecreasingSpatialTiles (tileLayout, gridBounds)
            elif layerName == "every-other-undefined":
                return EveryOtherSpatialTiles (tileLayout, gridBounds, float("nan"), 0.0)
            elif layerName == "every-other-0.99-else-1.01":
                return EveryOtherSpatialTiles (tileLayout, gridBounds, 0.99, 1.01)
            elif layerName == "every-other-1-else-1":
                return EveryOtherSpatialTiles (tileLayout, gridBounds, - 1, 1)
            elif layerName == "mod-10000":
                return ModSpatialTiles (tileLayout, gridBounds, 10000)
        spatialTestFile = generateSpatialTestFile()
        tiles = []
        for row in xrange(gridBounds.rowMin, gridBounds.rowMax+1):
            for col in xrange(gridBounds.colMin, gridBounds.colMax+1):
                key = SpatialKey(col, row)
                tile = spatialTestFile(key)
                tup = (key, tile)
                tiles.append(tup)
        return ContextRDD(sc.parallelize(tiles, _TestFiles.partitionCount), md)

    def spatialTestFile(self, name):
        return _TestFiles.generateSpatial(name)

    def spaceTimeTestFile(self, name):
        return _TestFiles.generateSpaceTime(name)

    def AllOnesTestFile(self):
        if self._AllOnesTestFile:
            return self._AllOnesTestFile
        self._AllOnesTestFile = self.spatialTestFile("all-ones")
        return self._AllOnesTestFile

    def AllTwosTestFile(self):
        if self._AllTwosTestFile:
            return self._AllTwosTestFile
        self._AllTwosTestFile = self.spatialTestFile("all-twos")
        return self._AllTwosTestFile

    def AllHundredsTestFile(self):
        if self._AllHundredsTestFile:
            return self._AllHundredsTestFile
        self._AllHundredsTestFile = self.spatialTestFile("all-hundreds")
        return self._AllHundredsTestFile

    def IncreasingTestFile(self):
        if self._IncreasingTestFile:
            return self._IncreasingTestFile
        self._IncreasingTestFile = self.spatialTestFile("increasing")
        return self._IncreasingTestFile

    def DecreasingTestFile(self):
        if self._DecreasingTestFile:
            return self._DecreasingTestFile
        self._DecreasingTestFile = self.spatialTestFile("decreasing")
        return self._DecreasingTestFile

    def EveryOtherUndefinedTestFile(self):
        if self._EveryOtherUndefinedTestFile:
            return self._EveryOtherUndefinedTestFile
        self._EveryOtherUndefinedTestFile = self.spatialTestFile("every-other-undefined")
        return self._EveryOtherUndefinedTestFile

    def EveryOther0Point99Else1Point01TestFile(self):
        if self._EveryOther0Point99Else1Point01TestFile:
            return self._EveryOther0Point99Else1Point01TestFile
        self._EveryOther0Point99Else1Point01TestFile = self.spatialTestFile("every-other-0.99-else-1.01")
        return self._EveryOther0Point99Else1Point01TestFile

    def EveryOther1ElseMinus1TestFile(self):
        if self._EveryOther1ElseMinus1TestFile:
            return self._EveryOther1ElseMinus1TestFile
        self._EveryOther1ElseMinus1TestFile = self.spatialTestFile("every-other-1-else-1")
        return self._EveryOther1ElseMinus1TestFile

    def Mod10000TestFile(self):
        if self._Mod10000TestFile:
            return self._Mod10000TestFile
        self._Mod10000TestFile = self.spatialTestFile("mod-10000")
        return self._Mod10000TestFile

    def AllOnesSpaceTime(self):
        if self._AllOnesSpaceTime:
            return self._AllOnesSpaceTime
        self._AllOnesSpaceTime = self.spaceTimeTestFile("spacetime-all-ones")
        return self._AllOnesSpaceTime

    def AllTwosSpaceTime(self):
        if self._AllTwosSpaceTime:
            return self._AllTwosSpaceTime
        self._AllTwosSpaceTime = self.spaceTimeTestFile("spacetime-all-twos")
        return self._AllTwosSpaceTime

    def AllHundredsSpaceTime(self):
        if self._AllHundredsSpaceTime:
            return self._AllHundredsSpaceTime
        self._AllHundredsSpaceTime = self.spaceTimeTestFile("spacetime-all-hundreds")
        return self._AllHundredsSpaceTime

    def CoordinateSpaceTime(self):
        if self._CoordinateSpaceTime:
            return self._CoordinateSpaceTime
        self._CoordinateSpaceTime = self.spaceTimeTestFile("spacetime-coordinates")
        return self._CoordinateSpaceTime
